### PR TITLE
[SHELL32] Implement 'Advanced' button of 'File Types'

### PR DESCRIPTION
**Pull request by katahiromz**
## Purpose

This PR allows the user to edit the File Type / extension information by clicking 'Advanced' button on 'File Types' property sheet of Folder Options.
JIRA issue: [CORE-12906](https://jira.reactos.org/browse/CORE-12906), [CORE-10972](https://jira.reactos.org/browse/CORE-10972)

## Proposed changes

- Add IDD_EDITTYPE and IDD_ACTION resource dialogs.
- Modify IDD_PICK_ICON resource dialog.
- Add NewActionDlgProc, EditActionDlgProc, and EditTypeDlgProc dialog procedures.
- Add 'ClassName' and 'IconLocation' members to FOLDER_FILE_TYPE_ENTRY structure.
- Implement IDC_FILETYPES_ADVANCED action on IDD_FOLDER_OPTIONS_FILETYPES.
- Add many resource strings.
- Fix and improve PickIconDlg.

### CODE REVIEW

**Comment by katahiromz** [comment]
2k3:
2k3-new-action
2k3-open-with
2k3-remove
2k3-use-dde
2ke-edit-action
2k3-change-icon
2k3-edit-type
2k3-error-1
2k3-error-2
2k3-error-3

**Comment by katahiromz** [comment]
ros:
ros-error-3
ros-new-action
ros-open-with
ros-remove
ros-change-icon
ros-edit-action
ros-edit-type
ros-error-1
ros-error-2

**Comment by Mouvedia** [comment]
Is the acronym DDE that well known? I think there's enough space to put _Dynamic Data Exchange_ instead.

**Comment by learn-more** [comment]
People that don't know `DDE` also won't know what it means to check the box `Dynamic Data Exchange`

File: dll/win32/shell32/dialogs/folder_options.cpp [hunk 1]
```diff
@@ -2144,6 +2151,708 @@ FileTypesDlg_OnItemChanging(HWND hwndDlg, PFOLDER_FILE_TYPE_ENTRY pEntry)
...
+{
+    // open class key
+    HKEY hKey;
+    if (RegOpenKeyExW(HKEY_CLASSES_ROOT, ClassKey, 0, KEY_READ, &hKey) != ERROR_SUCCESS)
+        return FALSE;
+
+    // open "shell" key
+    HKEY hShellKey;
+    if (RegOpenKeyExW(hKey, L"shell", 0, KEY_READ, &hShellKey) != ERROR_SUCCESS)
+        return FALSE;
```
[the remark this answers is not in the document]
Comment (katahiromz) [reply dll/win32/shell32/dialogs/folder_options.cpp]: OK.

File: dll/win32/shell32/dialogs/folder_options.cpp [hunk 2]
```diff
@@ -2144,6 +2151,708 @@ FileTypesDlg_OnItemChanging(HWND hwndDlg, PFOLDER_FILE_TYPE_ENTRY pEntry)
...
+            EnableWindow(GetDlgItem(hwndDlg, IDC_ACTION_USE_DDE), FALSE);
+            EnableWindow(GetDlgItem(hwndDlg, IDC_ACTION_ACTION), FALSE);
+            {
+                // set title
+                CStringW str(MAKEINTRESOURCEW(IDS_EDITING_ACTION));
+                str += s_pEditAct->ClassName;
+                SetWindowTextW(hwndDlg, str);
+            }
+            return TRUE;
+        case WM_COMMAND:
```
[the remark this answers is not in the document]
Comment (katahiromz) [reply dll/win32/shell32/dialogs/folder_options.cpp]: OK.

File: dll/win32/shell32/dialogs/folder_options.cpp [hunk 3]
```diff
@@ -2144,6 +2151,708 @@ FileTypesDlg_OnItemChanging(HWND hwndDlg, PFOLDER_FILE_TYPE_ENTRY pEntry)
...
+
+    // open class key
+    HKEY hClassKey;
+    if (RegOpenKeyExW(HKEY_CLASSES_ROOT, ClassKey, 0, KEY_WRITE, &hClassKey) != ERROR_SUCCESS)
+        return FALSE;
+
+    // open "shell" key
+    HKEY hShellKey;
+    if (RegOpenKeyExW(hClassKey, L"shell", 0, KEY_WRITE, &hShellKey) != ERROR_SUCCESS)
+        return FALSE;
```
[the remark this answers is not in the document]
Comment (katahiromz) [reply dll/win32/shell32/dialogs/folder_options.cpp]: OK.

File: dll/win32/shell32/dialogs/folder_options.cpp [hunk 4]
```diff
@@ -1460,14 +1462,17 @@ GetFileTypeIconsEx(PFOLDER_FILE_TYPE_ENTRY Entry, LPCWSTR IconLocation)
         Entry->hIconLarge = LoadIconW(shell32_hInstance, MAKEINTRESOURCEW(IDI_SHELL_EXE));
         Entry->hIconSmall = HICON(LoadImageW(shell32_hInstance, MAKEINTRESOURCEW(IDI_SHELL_EXE), IMAGE_ICON,
             GetSystemMetrics(SM_CXSMICON), GetSystemMetrics(SM_CYSMICON), 0));
+        StringCchCopyW(Entry->IconLocation, _countof(Entry->IconLocation), L"%SystemRoot%\\system32\\SHELL32.dll,-3");
     }
     else if (lstrcmpW(IconLocation, L"%1") == 0)
     {
+        StringCchCopyW(Entry->IconLocation, _countof(Entry->IconLocation), L"%SystemRoot%\\system32\\SHELL32.dll,-210");
```
[the remark this answers is not in the document]
Comment (katahiromz) [reply dll/win32/shell32/dialogs/folder_options.cpp]: It needs icon location string to pick the icon after this.

**Comment by katahiromz** [comment]
It seems like PickIconDlg won't work correctly. It doesn't show icons.

**Comment by katahiromz** [comment]
Done.

**Comment by katahiromz** [comment]
ros:
ros-no-icons
ros-icon-success
myext-test
myext-test-result

**Comment by katahiromz** [comment]
Done.

**Comment by HBelusca** [comment]
Hi! Can you split this PR in two? Namely:
- keep this PR for what it was originally for: the edition of "File Type" stuff;
- a second PR for just fixing the PickIconDlg function in shell32.

**Comment by katahiromz** [comment]
@HBelusca Both uses new resource strings. I'm not sure of possibility.

**Comment by HBelusca** [comment]
> @HBelusca Both uses new resource strings. I'm not sure of possibility. Please commit #574 at first.

OK, but you also need to remove the PickIconDlg-specific commits from this PR.

**Comment by katahiromz** [comment]
Can I use "git push -f"?

**Comment by HBelusca** [comment]
Not sure how that git command works, but I was more thinking about some:
git reset COMMIT_HASH_OF_PICKICONDLG_STUFF

For the corresponding commit hashes, just look at the latest commits you've done in this PR (e.g. "fix icon index handling" down to "fix PickIconDlg about icon index" ; however the "fix updating icon" commit seems to really belong to this PR instead so you should keep it here).

**Comment by katahiromz** [comment]
I had tried to resume the PickIconDlg-related changes but I'm not sure whether it conflicts or not.

File: dll/win32/shell32/dialogs/folder_options.cpp [hunk 5]
```diff
@@ -2199,13 +2963,37 @@ FolderOptionsFileTypesDlg(
...
+                    break;
             }
             break;
 
         case WM_NOTIFY:
             lppl = (LPNMLISTVIEW) lParam;
             switch (lppl->hdr.code)
             {
+                case LVN_KEYDOWN:
+                    {
```
Comment (HBelusca) [inline dll/win32/shell32/dialogs/folder_options.cpp]: Broken indentation. Braces should be aligned with 'c' of "case", and then the code block inside uses a 4-space indent.

**Comment by HBelusca** [comment]
In this PR the PickIconDlg "resume" has been done OK. Actually I was planning to merge this PR first and then the PickIconDlg one. In that latter one there will be certainly a minor conflict in the header file where the resource IDs are defined.

File: dll/win32/shell32/dialogs/folder_options.cpp [hunk 6]
```diff
@@ -2144,6 +2153,765 @@ FileTypesDlg_OnItemChanging(HWND hwndDlg, PFOLDER_FILE_TYPE_ENTRY pEntry)
...
+            // open 'New Action' dialog
+            if (IDOK == DialogBoxParamW(shell32_hInstance, MAKEINTRESOURCEW(IDD_ACTION), hwndDlg,
+                                        NewActionDlgProc, LPARAM(&action)))
+            {
+                if (SendMessageW(action.hwndLB, LB_FINDSTRING, -1, (LPARAM)action.szAction) != LB_ERR)
+                {
+                    // already exists, error
+                    CStringW strFormat(MAKEINTRESOURCEW(IDS_ACTION_EXISTS));
+                    TCHAR szText[256];
+                    StringCchPrintfW(szText, _countof(szText), strFormat, action.szAction);
```
[the remark this answers is not in the document]
Comment (katahiromz) [reply dll/win32/shell32/dialogs/folder_options.cpp]: Done.

**Pull request merged**